pkgs/tools/package-management/nix/common-meson.nix

```diff
@@ -0,0 +1,284 @@
+  suffix ? "",
+  hash ? null,
+  src ? fetchFromGitHub {
+    owner = "NixOS";
+    repo = "nix";
+    rev = version;
+    inherit hash;
+  },
+  patches ? [ ],
+  maintainers ? lib.teams.nix.members ++ [
+    lib.maintainers.lovesegfault
+    lib.maintainers.artturin
+  ],
+  self_attribute_name,
+}@args:
+assert (hash == null) -> (src != null);
+{
+  stdenv,
+  bison,
+  boehmgc,
+  boost,
+  brotli,
+  busybox-sandbox-shell,
+  bzip2,
+  callPackage,
+  cmake,
+  curl,
+  darwin,
+  doxygen,
+  editline,
+  flex,
+  git,
+  gtest,
+  jq,
+  lib,
+  libarchive,
+  libblake3,
+  libcpuid,
+  libgit2,
+  libsodium,
+  lowdown,
+  lowdown-unsandboxed,
+  toml11,
+  man,
+  meson,
+  ninja,
+  mdbook,
+  mdbook-linkcheck,
+  nlohmann_json,
+  nixosTests,
+  openssl,
+  perl,
+  python3,
+  pkg-config,
+  rapidcheck,
+  rsync,
+  Security,
+  sqlite,
+  util-linuxMinimal,
+  xz,
+  enableDocumentation ? stdenv.buildPlatform.canExecute stdenv.hostPlatform,
+  enableStatic ? stdenv.hostPlatform.isStatic,
+  withAWS ? !enableStatic && (stdenv.hostPlatform.isLinux || stdenv.hostPlatform.isDarwin),
+  aws-sdk-cpp,
+  withLibseccomp ? lib.meta.availableOn stdenv.hostPlatform libseccomp,
+  libseccomp,
+
+  confDir,
+  stateDir,
+  storeDir,
+
+  # passthru tests
+  pkgsi686Linux,
+  pkgsStatic,
+  runCommand,
+  pkgs,
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "nix";
+
+  version = "${version}${suffix}";
+  VERSION_SUFFIX = suffix;
+
+  inherit src patches;
+
+  outputs =
+    [
+      "out"
+      "dev"
+    ]
+    ++ lib.optionals enableDocumentation [
+      "man"
+      "doc"
+    ];
+
+  hardeningEnable = lib.optionals (!stdenv.hostPlatform.isDarwin) [ "pie" ];
+
+  hardeningDisable = [
+    "shadowstack"
+  ] ++ lib.optional stdenv.hostPlatform.isMusl "fortify";
+
+  nativeCheckInputs = [
+    git
+    man
+  ];
+
+  nativeBuildInputs =
+    [
+      bison
+      cmake
+      flex
+      jq
+      meson
+      ninja
+      pkg-config
+      rsync
+    ]
+    ++ lib.optionals enableDocumentation [
+      (lib.getBin lowdown-unsandboxed)
+      mdbook
+      mdbook-linkcheck
+    ]
+    ++ lib.optionals stdenv.hostPlatform.isLinux [
+      util-linuxMinimal
+    ];
+
+  buildInputs =
+    [
+      boost
+      brotli
+      bzip2
+      curl
+      editline
+      gtest
+      libarchive
+      libgit2
+      libsodium
+      lowdown
+      openssl
+      rapidcheck
+      sqlite
+      toml11
+      xz
+    ]
+    ++ lib.optionals (lib.versionAtLeast version "2.26") [
+      libblake3
+    ]
+    ++ lib.optionals enableDocumentation [
+      python3
+      doxygen
+    ]
+    ++ lib.optionals stdenv.hostPlatform.isDarwin [
+      Security
+    ]
+    ++ lib.optionals stdenv.hostPlatform.isx86_64 [
+      libcpuid
+    ]
+    ++ lib.optionals withLibseccomp [
+      libseccomp
+    ]
+    ++ lib.optionals withAWS [
+      aws-sdk-cpp
+    ]
+    ++ lib.optional (stdenv.hostPlatform.isDarwin) [
+      darwin.apple_sdk.libs.sandbox
+    ];
```

Comment:
Many of these things need to be propagated. The thing to check is whether they are a pkg-config public or pkg-config private dep. Public <=> propagated.